### PR TITLE
Added max reward checks in stake and redeem functions

### PR DESCRIPTION
**Pull request by deepesh-kn**
- Added checks in `EIP20Gateway.stake()` and `EIP20CoGateway.redeem()`.
- Check added is `amount > gasPrice * gasLimit`.
- Updated the test cases.

fixes: #568

### CODE REVIEW

File: contracts/gateway/EIP20CoGateway.sol [hunk 1]
```diff
@@ -867,11 +867,17 @@ contract EIP20CoGateway is GatewayBase {
...
             "Payable amount should be equal to the bounty amount."
         );
+
         require(
             _amount > uint256(0),
             "Redeem amount must not be zero."
         );
 
+        require(
+            _amount > _gasPrice.mul(_gasLimit),
```
Comment (schemar) [inline contracts/gateway/EIP20CoGateway.sol]: For readability we could replace `_gasPrice.mul(_gasLimit)` with a variable that has a meaningful name, e.g. `maxPossibleReward`. However, in solidity land I am not sure how to balance readability and cost here...

File: test/gateway/eip20_cogateway/redeem.js [hunk 2]
```diff
@@ -134,6 +135,27 @@ contract('EIP20CoGateway.redeem() ', function (accounts) {
     );
   });
 
+  it('should fail to max reward amount is greater than the redeem amount', async function () {
```
Comment (schemar) [inline test/gateway/eip20_cogateway/redeem.js]:
```suggestion
  it('should fail when max reward amount is greater than the redeem amount', async function () {
```

File: test/gateway/eip20_cogateway/redeem.js [hunk 3]
```diff
@@ -191,11 +213,11 @@ contract('EIP20CoGateway.redeem() ', function (accounts) {
 
   it('should fail when cogateway is not approved with redeem amount', async function () {
 
-    amount = new BN(100000);
+    let redeemAmount = amount.addn(1);
```
Comment (schemar) [inline test/gateway/eip20_cogateway/redeem.js]: Why? What's happening? 😕

File: test/gateway/eip20_cogateway/redeem.js [hunk 4]
```diff
@@ -227,17 +249,17 @@ contract('EIP20CoGateway.redeem() ', function (accounts) {
 
   it('should fail when the redeemer\'s BT balance is less than the redeem amount', async function () {
 
-    let amount = new BN(10000);
+    let redeemAmount = amount.muln(2);
```
Comment (schemar) [inline test/gateway/eip20_cogateway/redeem.js]: Why not `= balance.addn(1)` where `balance` is the redeemer's BT balance?

File: test/gateway/eip20_cogateway/redeem.js [hunk 5]
```diff
@@ -486,7 +508,7 @@ contract('EIP20CoGateway.redeem() ', function (accounts) {
     assert.strictEqual(
       (await utilityToken.balanceOf(redeemer)).eq(expectedBalance),
       true,
-      "Redeemer EIP20 token balance should be equal to ${expectedBalance}",
+      `Redeemer EIP20 token balance ${await utilityToken} should be equal to ${expectedBalance}`,
```
Comment (schemar) [inline test/gateway/eip20_cogateway/redeem.js]:
```suggestion
      `Redeemer's EIP20 token balance ${await utilityToken} should be equal to ${expectedBalance}`,
```

File: test/gateway/eip20_gateway/stake.js [hunk 6]
```diff
@@ -156,18 +156,24 @@ contract('EIP20Gateway.stake() ', function (accounts) {
...
 
   it('should fail to stake when beneficiary address is 0', async function () {
     beneficiary = Utils.NULL_ADDRESS;
-    errorMessage = "Beneficiary address must not be zero";
+    errorMessage = "Beneficiary address must not be zero.";
     await prepareData();
     await stake(utils.ResultType.FAIL);
   });
 
+  it('should fail to max reward amount is greater than the stake amount', async function () {
```
Comment (schemar) [inline test/gateway/eip20_gateway/stake.js]:
```suggestion
  it('should fail when max reward amount is greater than the stake amount', async function () {
```

File: test/gateway/eip20_gateway/stake.js [hunk 7]
```diff
@@ -156,18 +156,24 @@ contract('EIP20Gateway.stake() ', function (accounts) {
...
+    stakeAmount = new BN(200);
```
Comment (schemar) [inline test/gateway/eip20_gateway/stake.js]: What's the max reward amount here? Seems a little opaque.

File: test/gateway/eip20_cogateway/redeem.js (hunk 3 above)
Comment (deepesh-kn) [reply test/gateway/eip20_cogateway/redeem.js]: Added comment in the file.
```
    /*
     * CoGateway is approved to spend the redeem amount in beforeEach, so by
     * adding 1 to the approved(redeem) amount, the transfer will fail for this
     * test case.
     */
```

File: test/gateway/eip20_cogateway/redeem.js (hunk 4 above)
Comment (deepesh-kn) [reply test/gateway/eip20_cogateway/redeem.js]: done 👍

File: test/gateway/eip20_gateway/stake.js (hunk 7 above)
Comment (deepesh-kn) [reply test/gateway/eip20_gateway/stake.js]: Added comment in the file
```
    /*
     * Max reward amount will be `gasPrice * gasLimit`.
     * i.e. in this case its 180000000, which is higher than the stake amount.
     */
```

File: contracts/gateway/EIP20CoGateway.sol (hunk 1 above)
Comment (deepesh-kn) [reply contracts/gateway/EIP20CoGateway.sol]: I would prefer to keep like this. 
I will add a comment in the code. 
We always struggled a lot with stack too deep issue in this function.

Comment (deepesh-kn) [reply contracts/gateway/EIP20CoGateway.sol]: Added a comment in both stake and redeem functions.

Comment (schemar) [reply contracts/gateway/EIP20CoGateway.sol]: 👍

File: contracts/gateway/EIP20CoGateway.sol [hunk 8]
```diff
@@ -867,11 +867,22 @@ contract EIP20CoGateway is GatewayBase {
...
         );
 
+        /*
+         * Maximum reward possible is _gasPrice * _gasLimit, we check this
+         * upfront in this function to make sure that after unstake of the
+         * tokens it is possible to give the reward to the facilitator.
+         */
+        require(
+            _amount > _gasPrice.mul(_gasLimit),
+            "Maximum possible reward must be less than then the redeem amount."
```
Comment (sarvesh-ost) [inline contracts/gateway/EIP20CoGateway.sol]:
```suggestion
            "Maximum possible reward must be less than the redeem amount."
```

File: contracts/gateway/EIP20Gateway.sol [hunk 9]
```diff
@@ -304,11 +304,22 @@ contract EIP20Gateway is GatewayBase {
...
+        );
+
+        /*
+         * Maximum reward possible is _gasPrice * _gasLimit, we check this
+         * upfront in this function to make sure that after minting of the
+         * tokens it is possible to give the reward to the facilitator.
+         */
+        require(
+            _amount > _gasPrice.mul(_gasLimit),
+            "Maximum possible reward must be less than then the stake amount."
```
Comment (sarvesh-ost) [inline contracts/gateway/EIP20Gateway.sol]:
```suggestion
            "Maximum possible reward must be less than the stake amount."
```

File: test/gateway/eip20_gateway/stake.js [hunk 10]
```diff
@@ -156,18 +156,28 @@ contract('EIP20Gateway.stake() ', function (accounts) {
...
   });
 
+  it('should fail when max reward amount is greater than the stake amount', async function () {
+    /*
+     * Max reward amount will be `gasPrice * gasLimit`.
+     * i.e. in this case its 180000000, which is higher than the stake amount.
+     */
+    stakeAmount = new BN(200);
+    await prepareData();
+    errorMessage = "Maximum possible reward must be less than then the stake amount.";
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/stake.js]:
```suggestion
    errorMessage = "Maximum possible reward must be less than the stake amount.";
```

**Pull request merged**